### PR TITLE
(PIE-244) Add tasks to communicate with SN

### PR DESCRIPTION
**Pull request by gsparks**
This change renames the existing snow_record module to servicenow_tasks,
and simplifies the avaliable tasks to standard CRUD operations through
the ServiceNow Table API.

The rename is intented to communicate that this module does not install
ServiceNow, only serves as a place to store supporting tasks and plans
related to ServiceNow.

The README has also been simplified by moving examples to task metadata.

There will be a follow up PR to add the ability to query ServiceNow
records by attributes other than their sys_ids.

### CODE REVIEW

**Comment by ekinanp** [comment]
Since the tasks act on records, I think the naming should reflect that. E.g. `create` => `create_record`, `delete` => `delete_record`, etc. That way if I read these tasks in a plan's code, I'll immediately know the specific kind of thing they're acting on.

File: tasks/create.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
```
Comment (ekinanp) [inline tasks/create.json:1]: Let's change the `data` param to `fields`. Latter matches ServiceNow's docs. See "Request body"'s docs for `POST /now/table/{tableName}` in https://docs.servicenow.com/bundle/orlando-application-development/page/integrate/inbound-rest/concept/c_TableAPI.html#c_TableAPI

File: plans/example.pp [hunk 2]
```diff
@@ -1,31 +1,31 @@
-plan snow_record::example(
+plan servicenow_tasks::example(
```
Comment (ekinanp) [inline plans/example.pp:1]: Is this file necessary? I think we can assume that people using this module have a pretty good idea of how to write a plan using these tasks (that's unless `plans/example.pp` is a common convention for multiple modules, of which I don't know the answer to).

File: metadata.json [hunk 3]
```diff
@@ -1,12 +1,12 @@
 {
-  "name": "beersy-snow_record",
-  "version": "0.1.5",
-  "author": "Abir Majumdar",
+  "name": "puppetlabs-servicenow_tasks",
```
Comment (ekinanp) [inline metadata.json:2]: Huh, is this how the other `puppetlabs-<module_name>` modules do their metadata?

File: tasks/create.json (hunk 1 above)
Comment (ekinanp) [reply tasks/create.json:1]: Ditto for the `update` task.

File: tasks/create.rb [hunk 4]
```diff
@@ -8,13 +8,12 @@
 
 require_relative '../../ruby_task_helper/files/task_helper.rb'
 
-# This task creates objects
-class SnowCreate < TaskHelper
-  def task(table: 'incident',
+# This task creates records
+class ServiceNowCreate < TaskHelper
```
Comment (ekinanp) [inline tasks/create.rb:12]: We're definitely gonna need some better error handling here. Right now, this code swallows any API errors (because it doesn't check for the response's status, it just prints it as-is). It'd also be nice to DRY up some of this common "perform request" code, and avoid having to extract credentials all the time. That "perform request" code could do the error-checking.

Comment (gsparks) [reply tasks/create.rb:12]: I'm wondering if we want all responses (200 or not) to go to stdout since the task did it's job to get that response properly. Then we would print any errors to stderr that actually happened inside the task.

File: metadata.json (hunk 3 above)
Comment (gsparks) [reply metadata.json:2]: I'll check again, but I think so.

File: plans/example.pp (hunk 2 above)
Comment (gsparks) [reply plans/example.pp:1]: I think it's common to have examples, but typically they are in an examples directory.

 examples/
    Contains examples showing how to declare the module's classes and defined types.
    init.pp: The main class of the module.
    example.pp: Provide examples for major use cases.

Comment (ekinanp) [reply plans/example.pp:1]: Cool. Let's do that then?

Comment (gsparks) [reply plans/example.pp:1]: I moved it to examples, but you can't actually run it from there. Also it breaks rubocop. Moved it back to plans.

File: tasks/create.rb (hunk 4 above)
Comment (gsparks) [reply tasks/create.rb]: Moved the http code to a separate class, but still thinking about the best way to handle non 200s. Sent errors to STDERR.

File: metadata.json (hunk 3 above)
Comment (gsparks) [reply metadata.json:2]: This looks consistent with the modules I looked at.

File: README.md [hunk 5]
```diff
@@ -1,43 +1,29 @@
...
-* create - Create a ServiceNow object
-* read - Get a ServiceNow object
-* update - Update a ServiceNow object
-
-There are also 2 additional tasks for interacting with Incidents.
+servicenow_tasks provides a series of tasks to interact with a ServiceNow Instance. The 4 main tasks included with this module:
 
-* create_incident - Create an Incident
-* resolve_incident - Resolve an Incident
+* create_record - Create a ServiceNow record
```
Comment (ekinanp) [inline README.md:15]: Is listing these necessary? Couldn't they see them via `bolt task show`? Thinking in the future if we add new tasks, it might be annoying to have to remember to update a README.

File: lib/service_now_request.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+require 'json'
+require 'net/http'
+require 'openssl'
+
+# This class organizes ruby http code used by all of the tasks.
+class ServiceNowRequest
+  def initialize(uri, http_verb, fields, user, password)
+    @uri = URI.parse(uri)
+    @http_verb = http_verb.capitalize
+    @fields = fields
```
Comment (ekinanp) [inline lib/service_now_request.rb:11]: I think here, `fields` => `body`. `ServiceNowRequest` is more general and could apply to requests outside of the table API.

File: metadata.json (hunk 3 above)
Comment (ekinanp) [reply metadata.json:2]: Ah ok. How hard would it be to change the repo name to `puppetlabs-servicenow_tasks` ? Just feels weird having that naming inconsistency compared w/ the other repos.

File: plans/example.pp (hunk 2 above)
Comment (ekinanp) [reply plans/example.pp:1]: That's annoying because this will show up in `bolt plan show`. Could we maybe add this plan to the README instead, like under an `Example Plan` section?

File: tasks/create_record.json [hunk 7]
```diff
@@ -1,16 +1,16 @@
 {
     "puppet_task_version": 1,
     "supports_noop": false,
-    "description": "Create a ServiceNow object",
+    "description": "Create a ServiceNow record\nbolt task run --nodes dev84270 servicenow_tasks::create_record table=incident fields='{\"description\":\"An example incident.\"}'",
```
Comment (ekinanp) [inline tasks/create_record.json:4]: What does `bolt task show` output for this task? What does `bolt task show <task>` look like? Curious to see if the multi-line description does anything weird in the output.

File: tasks/create_record.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    user = _target[:user]
+    password = _target[:password]
+    instance = _target[:name]
+
+    uri = "https://#{instance}.service-now.com/api/now/table/#{table}"
+
+    begin
+      request = ServiceNowRequest.new(uri, 'Post', fields, user, password)
+      request.print_response
+    rescue => e
```
Comment (ekinanp) [inline tasks/create_record.rb:22]: Since the only possible errors are from the request, I think the `begin/end` block could move over to `print_response`. Then you can have in `print_response`
```
begin
rescue => e
  raise TaskHelper::Error...
end
```

Note that a Ruby exception will always print to stderr, and that we will print the right thing because `TaskHelper::Error` handles that. So, having `STDERR.puts` is redundant.

File: tasks/read_record.rb [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    msg: message,
+    kind: 'servicenow_tasks.read_record',
+    details: {},
+  }
+  # puts result.to_json
+  raise TaskHelper::Error.new('Failure!', 'snow_record.read_record', result)
+end
+
+# This task reads incidents
+class ServiceNowRead < TaskHelper
```
Comment (ekinanp) [inline tasks/read_record.rb:19]: This could be a nit, but I think `read_record` => `get_record`. I think of `read_<thing>` as "read `<thing>`'s content". Records don't have "content" (not like files). They're just a JSON of fields, which is what this is returning, and that "JSON of fields" is better captured by `get_record`.

File: tasks/create_record.json (hunk 7 above)
Comment (gsparks) [reply tasks/create_record.json:4]: I messed around with a few different ways to format it, and this one isn't great, but it's better than having it all on one line.

File: lib/service_now_request.rb (hunk 6 above)
Comment (ekinanp) [reply lib/service_now_request.rb:11]: Oh also, should this be in the module's `files` directory? Asking b/c that's where the task helper module puts its code as well.

File: plans/example.pp (hunk 2 above)
Comment (gsparks) [reply plans/example.pp]: That seems cluttery too. I think your original instinct was right. This doesn't need to be the place to look for how to write a plan. I'm just deleting it.

File: metadata.json (hunk 3 above)
Comment (gsparks) [reply metadata.json:2]: Sorry, I think I misled you. Other modules have the same setup as we do in the metadata where the name would be puppetlabs-servicenow_tasks, the repo name is servicenow_tasks, and the repo namespace is puppetlabs. So I think the way it is now is consistent with the rest of the modules up on the forge.

File: README.md (hunk 5 above)
Comment (gsparks) [reply README.md:15]: It's only four things and it could be nice to see whats available from the README without installing the module. Just my 2 cents.

File: lib/service_now_request.rb (hunk 6 above)
Comment (gsparks) [reply lib/service_now_request.rb]: I think according to this [doc](https://puppet.com/docs/puppet/6.10/modules_fundamentals.html#concept-1234) lib sounds slighty better, but I'm open to changing if you disagree.

File: README.md (hunk 5 above)
Comment (ekinanp) [reply README.md:15]: Yeah, I don't have strong feelings. I just personally hate updating stuff unless I really have to.

File: lib/service_now_request.rb (hunk 6 above)
Comment (gsparks) [reply lib/service_now_request.rb]: Change `fields` to `body` for the ServiceNowRequest class.

File: metadata.json (hunk 3 above)
Comment (ekinanp) [reply metadata.json:2]: Ah ok. Thanks for clarifying that.

File: lib/service_now_request.rb (hunk 6 above)
Comment (ekinanp) [reply lib/service_now_request.rb]: Ohh my bad, I just noticed this is in `lib`. No problem, thanks.

File: tasks/create_record.rb (hunk 8 above)
Comment (gsparks) [reply tasks/create_record.rb:22]: Standardized the error handling. Removed redundant STDERR.puts.

File: tasks/read_record.rb (hunk 9 above)
Comment (gsparks) [reply tasks/read_record.rb]: Changed name to get_record.

File: tasks/create_record.rb (hunk 8 above)
Comment (gsparks) [reply tasks/create_record.rb]: moved begin block to ServiceNowRequest.rb

File: lib/service_now_request.rb [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        # Make request to ServiceNow
+        response = http.request(request)
+
+        # Parse and print response
+        json_response = JSON.parse(response.body)
+        pretty_response = JSON.pretty_unparse(json_response)
+        puts [pretty_response]
+      end
+    rescue => e
+      puts "ERROR: #{e}"
```
Comment (ekinanp) [inline lib/service_now_request.rb:48]: Ah, I meant that when you `raise <Exception>`, that automatically gets printed to stderr. So you don't need the `puts` there. In the original comment, I meant that you could delete the `STDERR.puts` line.

Comment (gsparks) [reply lib/service_now_request.rb:48]: ahh got it

Comment (gsparks) [reply lib/service_now_request.rb]: fixed

File: tasks/create_record.json [hunk 11]
```diff
@@ -1,16 +1,16 @@
...
     "parameters": {
         "table": {
             "description": "ServiceNow table",
             "type": "Optional[String[1]]"
         },
-        "data": {
-            "description": "ServiceNow data",
+        "fields": {
+            "description": "ServiceNow fields",
             "type": "String[1]"
```
Comment (ekinanp) [inline tasks/create_record.json:14]: Ah just noticed that this is a String. This should probably be a hash-type of <field> => <value>. You could probably update the description to include that.

Comment (ekinanp) [reply tasks/create_record.json:14]: Also, I think the description for fields could be `ServiceNow fields. This is a hash of <field_name> => <value>`.

File: tasks/create_record.json (hunk 7 above)
Comment (ekinanp) [reply tasks/create_record.json:4]: Yeah, looks like `bolt task show <task_name>` outputs a stub usage. So we probably don't need the `bolt task run` part.

File: tasks/delete_record.json [hunk 12]
```diff
@@ -1,21 +1,17 @@
...
             "type": "Optional[String[1]]"
         },
         "sys_id": {
             "description": "ServiceNow sys_id",
-            "type": "String[1]"
-        },
-        "data": {
-            "description": "ServiceNow data",
-            "type": "String[1]"
+            "type": "Optional[String[1]]"
```
Comment (ekinanp) [inline tasks/delete_record.json:14]: I don't think this should be optional.

File: tasks/get_record.json [hunk 13]
```diff
@@ -1,20 +1,16 @@
...
     "puppet_task_version": 1,
     "supports_noop": false,
-    "description": "Read a ServiceNow object",
+    "description": "Get a ServiceNow record\nbolt task run --nodes dev84270 servicenow_tasks::get_record table=incident fields='{\"sysid\"=\"3Db9b91557db00101096dde37a48961976\"}'",
     "remote": true,
     "files": ["ruby_task_helper/files/task_helper.rb"],
     "parameters": {
         "table": {
             "description": "ServiceNow table",
             "type": "Optional[String[1]]"
```
Comment (ekinanp) [inline tasks/get_record.json:10]: This shouldn't be `Optional`.

File: tasks/get_record.json [hunk 14]
```diff
@@ -1,20 +1,16 @@
...
         },
-        "number": {
-            "description": "ServiceNow number",
-            "type": "Optional[String[1]]"
-        },
-        "lookup_field": {
-            "description": "ServiceNow lookup_field",
+        "fields": {
+            "description": "ServiceNow fields to query by",
             "type": "Optional[String[1]]"
```
Comment (ekinanp) [inline tasks/get_record.json:14]: This also shouldn't be `Optional`. Also, this should probably be a hash.

Comment (gsparks) [reply tasks/get_record.json]: made it mandatory hash.

File: tasks/get_record.json (hunk 13 above)
Comment (gsparks) [reply tasks/get_record.json]: fixed

File: tasks/delete_record.json (hunk 12 above)
Comment (gsparks) [reply tasks/delete_record.json]: fixed

File: tasks/create_record.json (hunk 11 above)
Comment (gsparks) [reply tasks/create_record.json]: fixed

File: tasks/create_record.json [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+    "supports_noop": false,
+    "description": "Create a ServiceNow record",
+    "files": ["ruby_task_helper/files/task_helper.rb"],
+    "remote": true,
+    "parameters": {
+        "table": {
+            "description": "ServiceNow table",
+            "type": "String[1]"
+        },
+        "servicenow_fields": {
```
Comment (ekinanp) [inline tasks/create_record.json:12]: Why is this `servicenow_fields` instead of `fields` ? Also, I think description should be "The record's fields. This is a hash of <field_name> => <field_value>" (the `<field_name>`, `<field_value>` stuff makes the param type's meaning clearer imo).

File: tasks/create_record.json [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+            "description": "ServiceNow fields",
+            "type": "Hash[String, String]"
```
Comment (ekinanp) [inline tasks/create_record.json:14]: Hmm, I think we'll want the value-type to be Puppet's equivalent of `Any Type` instead of `String` since ServiceNow fields can have arbitrary types (String, Integer, Boolean). Otherwise, if I have a numeric field `A` and I pass in `A=3` to the task, the API call will fail b/c the endpoint will receive it as `"A": "3"` instead of `"A": 3`.

File: tasks/create_record.json (hunk 15 above)
Comment (gsparks) [reply tasks/create_record.json]: Sorry I misunderstood your original comment, fixed.

File: tasks/create_record.json (hunk 16 above)
Comment (gsparks) [reply tasks/create_record.json]: fixed.

File: tasks/create_record.json [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+    "files": ["ruby_task_helper/files/task_helper.rb"],
+    "remote": true,
+    "parameters": {
+        "table": {
+            "description": "ServiceNow table",
+            "type": "String[1]"
+        },
+        "fields": {
+            "description": "ServiceNow fields. This is a hash of <field_name> => <value>",
+            "type": "Hash[Any, Any, 1]"
```
Comment (ekinanp) [inline tasks/create_record.json:14]: Yeah, I guess creating a record with no fields would be weird. Good call.

File: tasks/get_record.json [hunk 18]
```diff
@@ -1,17 +1,17 @@
...
         "table": {
             "description": "ServiceNow table",
-            "type": "Optional[String[1]]"
-        },
-        "data": {
-            "description": "ServiceNow data",
             "type": "String[1]"
+        },
+        "fields": {
+            "description": "ServiceNow fields. This is a hash of <field_name> => <value>",
```
Comment (ekinanp) [inline tasks/get_record.json:13]: So what is `fields` here? Are those the fields to filter on when retrieving record? Would be nice to provide a bit more detail

File: tasks/update_record.json [hunk 19]
```diff
@@ -1,21 +1,21 @@
...
-            "description": "ServiceNow close notes",
+        "sys_id": {
+            "description": "ServiceNow sys_id",
             "type": "String[1]"
         },
-        "additional_data": {
-            "description": "ServiceNow additional data",
-            "type": "String[1]"
+        "fields": {
+            "description": "ServiceNow fields. This is a hash of <field_name> => <value>",
```
Comment (ekinanp) [inline tasks/update_record.json:17]: I'd maybe change "ServiceNow fields ..." => "The updated fields. ...." Latter provides a bit more context.

File: tasks/get_record.json (hunk 18 above)
Comment (ekinanp) [reply tasks/get_record.json:13]: Also, I think maybe for this one, `fields` could be Optional. Seems weird to have to specify `fields` when getting a record.

Comment (gsparks) [reply tasks/get_record.json:13]: yeah in this case the fields are what gets used as a query on the table. If you don't specify anything I think you just get the entire table.

Comment (gsparks) [reply tasks/get_record.json]: changed to query_fields to be more clear
